features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/domain/ResourceRequestBodyFactory.kt

```diff
@@ -0,0 +1,140 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.domain
+
+import com.datadog.android.api.storage.RawBatchEvent
+import com.datadog.android.sessionreplay.internal.exception.InvalidPayloadFormatException
+import com.datadog.android.sessionreplay.internal.utils.MiscUtils
+import com.google.gson.JsonObject
+import okhttp3.MediaType.Companion.toMediaTypeOrNull
+import okhttp3.MultipartBody
+import okhttp3.RequestBody
+import okhttp3.RequestBody.Companion.toRequestBody
+import kotlin.jvm.Throws
+
+internal class ResourceRequestBodyFactory {
+
+    @Throws(InvalidPayloadFormatException::class)
+    internal fun create(
+        resources: List<RawBatchEvent>
+    ): RequestBody {
+        @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
+        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
+        addResourcesSection(builder, resources)
+        addApplicationIdSection(builder, resources)
+        @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
+        return builder.build()
+    }
+
+    private fun getApplicationId(resources: List<RawBatchEvent>): String {
+        val resourcesMetaData: List<JsonObject> = resources.mapNotNull {
+            MiscUtils.safeDeserializeToJsonObject(it.metadata)
+        }
+
+        val uniqueApplicationIds = mutableSetOf<String>()
+        resourcesMetaData.forEach {
+            val applicationId = it.get(APPLICATION_ID_KEY)?.asString
+
+            if (applicationId != null) {
+                uniqueApplicationIds.add(applicationId)
+            }
+        }
+
+        if (uniqueApplicationIds.size == 0) {
+            @Suppress("ThrowingInternalException")
+            throw InvalidPayloadFormatException(UNABLE_GET_APPLICATION_ID_ERROR)
+        }
+
+        if (uniqueApplicationIds.size > 1) {
+            @Suppress("ThrowingInternalException")
+            throw InvalidPayloadFormatException(MULTIPLE_APPLICATION_ID_ERROR)
```

Comment:
This is a bit of a problem, because it means that a list of resources with two different applicationIds won't send correctly. OTOH, I'm not sure we can prevalidate and split this before it reaches the factory

Reply:
I've changed to check for the number of applicationIds - if 0 then throw an invalidPayloadException, if greater than 1 then determine the most "popular" applicationId, filter the list of events by this id and send them in the request, and log the issue through the internalLogger